### ISSUE TITLE
Pydantic examples throw regex vocabulary errors

### ISSUE THREAD

**Issue by janvandervegt-db** [outlines-dev/outlines#337]
### Describe the issue as clearly as possible:

When running the `dating_profile` example (or for my own case), I'm getting a `ValueError: The vocabulary does not allow us to build a sequence that matches the input regex` issue.

### Steps/code to reproduce the bug:

```python
Direct copy from `examples/dating_profile.py`
```

### Expected result:

```shell
Properly formatted JSON strings.
```

### Error message:

```shell
121 new_description = "I'm a laid-back lawyer who spends a lot of his free-time gaming. I work in a corporate office, but ended up here after the start-up I cofounded got acquired, so still play ping pong with my cool coworkers every day. I have a bar at home where I make cocktails, which is great for entertaining friends. I secretly like to wear suits and get a new one tailored every few months. I also like weddings because I get to wear those suits, and it's a good excuse for a date. I watch the latest series because I'm paying, with my hard-earned money, for every streaming service."
    123 prompt = dating_profile_prompt(description=new_description, examples=samples)
--> 124 profile = text.generate.json(model, DatingProfile)(prompt)
    125 print(profile)
    127 parsed_profile = DatingProfile.model_validate_json(profile)

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-75b35321-4d37-4b53-b4f1-bbf66569a6ea/lib/python3.10/site-packages/outlines/text/generate/regex.py:424, in json(model, schema, max_tokens, sampler, allow_empty_tokens)
    420     schema = dumps(schema.model_json_schema())
    422 regex_str = build_regex_from_schema(schema)
--> 424 return Regex(
    425     model,
    426     regex_str,
    427     max_tokens,
    428     sampler=sampler,
    429     allow_empty_tokens=allow_empty_tokens,
    430 )

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-75b35321-4d37-4b53-b4f1-bbf66569a6ea/lib/python3.10/site-packages/outlines/text/generate/regex.py:107, in Regex.__init__(self, model, regex_string, max_tokens, sampler, stop, allow_empty_tokens, initial_state, final_states, states_to_token_maps, empty_token_ids)
    101 # Check whether a terminal path (from the initial state of the FSM to
    102 # one of its terminal states) exists, raise an exception otherwise.
    103 if not any(
    104     self.final_states.intersection(v.values())
    105     for v in self.states_to_token_maps.values()
    106 ):
--> 107     raise ValueError(
    108         "The vocabulary does not allow us to build a sequence that matches the input regex"
    109     )
    111 # When an EOS is observed, the last FSM state becomes `-1`.
    112 self.last_fsm_states: List[int] = []

ValueError: The vocabulary does not allow us to build a sequence that matches the input regex
```

### Outlines/Python version information:

```python
Running `outlines.__version__` throws an AttributeError for me, but `pip list` shows `0.0.9`.

Python version is 3.10.12

Pydantic is 2.4.2

pydantic-core is 2.10.1
```

### Context for the issue:

I really want to use Pydantic for structured generation but it's not working :)

**Comment by rlouf** [comment]
Thank you for opening an issue! Are you using the same model as the example's?

**Comment by janvandervegt** [comment]
Yes, it is a straight up copy/paste of the example (on a Databricks notebook with a GPU). 

I did make some progress, I was having issues with numpy `1.25.x`, so I downgraded it to `1.21.x` which is what I was running the example with. Upgrading to `1.26.x` seemed to have solved the issue (although it is extremely slow, I don't have a reference point for expected runtime).

**Comment by rlouf** [comment]
> although it is extremely slow, I don't have a reference point for expected runtime)

Which version of Outlines are you using?

**Comment by janvandervegt** [comment]
I'm using version `0.0.9`, the Pydantic model is small but nested:

```
class Column(BaseModel):
    name: str
    data_type: str

class Table(BaseModel):
    name: str
    database: str
    catalog: str
    columns: List[Column]

class Tables(BaseModel):
    tables: List[Table]
```

Running the generation without outlines takes ~10 seconds, running it with outlines on a single prompt takes ~45 seconds, running it on a list of two prompts takes around 5 minutes.

**Comment by rlouf** [comment]
Interestingly running it on my machine only results in a 9s delta. These 9 seconds comes from the compilation of the Finite State Machines that guides the generation. You only have to perform this compilation once during the same session as it is cached. 

Take the following code:

```python
from datetime import datetime
import json
from enum import Enum
from typing import List

from pydantic import BaseModel, constr
import interegular

import outlines.models as models
import outlines.text.generate as generate
from outlines.text.json_schema import build_regex_from_schema
from outlines.text.fsm import make_deterministic_fsm
from outlines.text.fsm import create_fsm_index_tokenizer

## Setup the Pydantic model and load GPT2
class Column(BaseModel):
    name: str
    data_type: str

class Table(BaseModel):
    name: str
    database: str
    catalog: str
    columns: List[Column]

class Tables(BaseModel):
    tables: List[Table]

model = models.transformers("gpt2")



# Let us compile the finite state machine that guides the generation
start = datetime.now()
regex_str = build_regex_from_schema(json.dumps(Tables.model_json_schema()))
regex_pattern = interegular.parse_pattern(regex_str)
regex_fsm, _ = make_deterministic_fsm(regex_pattern.to_fsm().reduce())
_ = create_fsm_index_tokenizer(regex_fsm, model.tokenizer)
now = datetime.now()

print(f"{now-start}")
# 00:00:09

# Let us compile it a second time. It should be cached
now = datetime.now()
regex_str = build_regex_from_schema(json.dumps(Tables.model_json_schema()))
regex_pattern = interegular.parse_pattern(regex_str)
regex_fsm, _ = make_deterministic_fsm(regex_pattern.to_fsm().reduce())
_ = create_fsm_index_tokenizer(regex_fsm, model.tokenizer)
last = datetime.now()

print(f"{last-now}")
# 00:00:03
```

I see three things here:
1. Caching is not done at the right place; the second compilation should take a few miliseconds at most (it's a cache query)
2. We could cache the compilation artifacts to disk so they persist between session;
3. It seems that somehow batching is highly inefficient in outlines.

I will need to check (3) on my machine, and open issues accordingly.

**Comment by janvandervegt-db** [comment]
I've ran some of my own tests and it looks like the batching observation was just an outlier, though my suspicion based on the number of tokens generated is that it generates them in sequence instead of in parallel. Is that expected?

I also sometimes run into an assertion error when using batches, with line 156 in `text/generate/regex.py`:

`assert last_state > -1`

This is using `mosaicml/mpt-7b-instruct` instead of `gpt-2` though.

Let me know if you want me to open a different issue for this.

**Comment by rlouf** [comment]
> I've ran some of my own tests and it looks like the batching observation was just an outlier, though my suspicion based on the number of tokens generated is that it generates them in sequence instead of in parallel. Is that expected?

The only thing that runs sequentially is the constraining step; the forward pass through the model happens in parallel afair.

> I also sometimes run into an assertion error when using batches, with line 156 in `text/generate/regex.py`:
> 
> `assert last_state > -1`
> 
> This is using `mosaicml/mpt-7b-instruct` instead of `gpt-2` though.
> 
> Let me know if you want me to open a different issue for this.

It would be nice if you could open a new issue with an example to reproduce the issue (don;'t forget to set the seed!).

**Comment by janvandervegt-db** [comment]
Thanks so much, this library promises exactly what I was looking for - I'm going to do some more tests and happy to contribute to `outlines`.

Here is the issue related to the batching: https://github.com/outlines-dev/outlines/issues/340
